mlflow/catboost.py

```diff
@@ -146,13 +146,29 @@ def save_model(
     model_data_path = os.path.join(path, _MODEL_BINARY_FILE_NAME)
     cb_model.save_model(model_data_path, **kwargs)
 
-    conda_env, pip_requirements, pip_constraints = (
-        _process_pip_requirements(
-            get_default_pip_requirements(), pip_requirements, extra_pip_requirements,
+    flavor_conf = {
+        _MODEL_TYPE_KEY: cb_model.__class__.__name__,
+        _SAVE_FORMAT_KEY: kwargs.get("format", "cbm"),
+    }
+
+    if conda_env is None:
+        default_reqs = get_default_pip_requirements()
+        if pip_requirements is None:
+            # HACK: Temporarily create an MLmodel file because `mlflow.catboost._load_pyfunc`
+            # requires it to get model type and save format.
+            save_path = os.path.join(path, MLMODEL_FILE_NAME)
+            Model().add_flavor(FLAVOR_NAME, **flavor_conf).save(save_path)
+            inferred_reqs = mlflow.infer_pip_requirements(
+                model_data_path, FLAVOR_NAME, fallback=default_reqs,
+            )
+            os.remove(save_path)  # Clean up the MLmodel file
```

Comment:
Instead of writing the MLmodel file and then removing it, can we save `mlflow_model` to a file before this `if` block (right below the definition of `flavor_conf`) and then reload it after? Can we move the inline comment as well?

Can we also clarify in the comment that `mlflow.catboost._load_pyfunc` is called for dependency inference purposes? Otherwise, it's unclear why we care about `_load_pyfunc`.

Finally, can we avoid describing this as a hack? It's part of the model persistence procedure :)

Reply:
Got it. I misunderstood we need to save `conda_env.yml` and `requirements.txt` before `mlflow_model.save`, but we can save them after `mlflow_model.save`.